### DISCUSSION TITLE
question about thread safe

### DISCUSSION THREAD

**Discussion by calvin2021y** [duckdb/duckdb#4260] [Q&A] [answered]
This is my first time use duckdb.   because 0.4.0 some bugs, I use master branch.

I open 4 connection from same db object from 4 thread,  (db object create at app start),  if I do not append any new data, then the server give same results at all thread.

after I append or delete some data, then some thread is give old results.  (show deleted records, or missing new insert records)

Is this a bugs?

**Comment by hannes** [comment]
Can you try to make a reproducible example? In general, DuckDB uses MVCC, so its possible you see "old" data if you are in an uncommitted transaction. So you might try to COMMIT and see if the old/new records are as you expect.

**Reply by calvin2021y** [reply]
When I do insert, I create new appender and destroy it when I am done( ASAP ),  do I need do some more to commit it ?

I don`t see any code related to commit from https://duckdb.org/docs/api/c/appender

**Reply by Alex-Monahan** [reply]
Hello! 

Please follow the issue template. It makes it easier to help you. You are not giving us enough information to see the issue that you are having. Please provide all of the exact code and data needed to cause the problem. 

Thanks!

**Reply by calvin2021y** [reply]
So it is a bugs ?  If so I will report late when I reduce the example.

I ask because I am not sure, if there is  some thread sync function is not call ,  or I can not create thread connect for each of them

**Reply by Mytherin** [reply]
It sounds to me like you are not calling flush or destroy on the appender, or are not committing from one of the threads. It is impossible to tell unless you provide your code, however.

**Reply by calvin2021y** [reply]
If I am not flush it, there should be no correct results from all thread,  but some thread give correct results.

**Comment by calvin2021y** [comment]
the code like this:

```c
                doQuery("BEGIN TRANSACTION");
                if (duckdb_appender_create(conn, null, "orders", &appender) == DuckDBError) {
                    assert(false);
                }
            do_append without error;
            duckdb_appender_destroy(&appender);
            doQuery("COMMIT");
```

I also try adjust query order:

```c

                if (duckdb_appender_create(conn, null, "orders", &appender) == DuckDBError) {
                    assert(false);
                }
                doQuery("BEGIN TRANSACTION");
            do_append without error;
            doQuery("COMMIT");
            duckdb_appender_destroy(&appender);

```

I print into console when destroy it,  still get wrong results form other thread.

I also has try :

```c
        duckdb_set_config(config, "disabled_optimizers", "true");
         duckdb_set_config(config, "enable_object_cache", "false");
```

**Reply by Mytherin** [reply]
Could you please provide a full compilable example? This is not a full example.

**Reply by Mytherin** [reply]
Thanks, that is more helpful, albeit still not a compilable/complete example. You are sharing the same connection over multiple threads. That is not recommended, you should open a new connection per thread. You might also want to explicitly call `duckdb_appender_flush` to ensure the data is flushed.

**Reply by calvin2021y** [reply]
the projects is huge,  this is the insert code.

And there is libuv worker thread, each of them has when own conenction obejct with tls vars.    db object open before libuv event loop start.

**Reply by Mytherin** [reply]
Alright, well, we are not here to provide you free support on your huge project. Please create a reproducible example using only DuckDB code that illustrates a problem with DuckDB, which we can then act on and fix. If you are not able to do that then we cannot help you.

**Reply by calvin2021y** [reply]
I just add `duckdb_appender_flush(appender);` before destroy.  

```c
struct duckdb_connection {
    void* ___ptr;
}
```
I print when I create new connection ( duckdb_connection.___ptr),  every thread has they own connection object.

There alway only one append task , before one finish there is no new append task created.

still get the error results from other worker thread.

**Reply by calvin2021y** [reply]
after wrong reulsts,  stop app then restart it.  every thread get same results again.  (before restart, some give good results, some give old restuls.)

**Reply by calvin2021y** [reply]
> Alright, well, we are not here to provide you free support on your huge project. Please create a reproducible example using only DuckDB code that illustrates a problem with DuckDB, which we can then act on and fix. If you are not able to do that then we cannot help you.

I hope I will reduce it, before this I will try open new connection every time I need do a query.

**Comment by calvin2021y** [comment]
I wloud like to provde a minimal example, but it is not easy task since there is a lot basic code in D but not in C.

I fix this problem by open new connection every time and close it when I am finish the task(few query in a TRANSACTION), and  call duckdb_disconnect it ASAP.

**Comment by calvin2021y** [comment, marked as answer]
turn out this cause by a logic bugs( the dababase reopened without close)
